libass/ass_shaper.c

```diff
@@ -221,8 +221,13 @@ get_cached_metrics(struct ass_shaper_metrics_data *metrics,
     if (metrics->vertical && unicode >= VERTICAL_LOWER_BOUND)
         rotate = true;
 
-    metrics->hash_key.glyph_index = glyph;
-    FT_Glyph_Metrics *val = ass_cache_get(metrics->metrics_cache, &metrics->hash_key,
+    GlyphMetricsHashKey key = {
+        .font = metrics->hash_key.font,
+        .face_index = metrics->hash_key.face_index,
+        .size = metrics->hash_key.size,
+        .glyph_index = glyph,
+    };
+    FT_Glyph_Metrics *val = ass_cache_get(metrics->metrics_cache, &key,
```

Comment:
there are two `get_cached_metrics` calls in `get_glyph_nominal` and `get_glyph_variation` with a leading comment suggesting it might or at some point might have relied on side effects from this mutation:
```C
    // rotate glyph advances for @fonts while we still know the Unicode
codepoints
    get_cached_metrics(metrics_priv, unicode, *glyph);
```

They were added in 1636a551c7d7b05ea15c63a4ad041f2e8d8eb79f and dbb76629145567458c683823a1c20a3c6c2aae64 and respectively

Reply:
That's not about mutation of the cache key; it's about making sure that the first call to `get_cached_metrics` on a given event for a given key will always have the correct value for `unicode` (whereas later calls may pass 0). `unicode` is used to determine the value passed to the `priv` arg of `ass_cache_get`, and thus to `ass_glyph_metrics_construct` on cache miss (and misses can only happen on the first call of a frame, or in a future threaded world, the first of an event).